lib/kernel.js

```diff
@@ -16,34 +16,89 @@ import store from "./store";
 import WatchesStore from "./store/watches";
 import OutputStore from "./store/output";
 import HydrogenKernel from "./plugin-api/hydrogen-kernel";
+import InputView from "./input-view";
+import KernelTransport from "./kernel-transport";
+import ZMQKernel from "./zmq-kernel";
+import type { ResultsCallback } from "./kernel-transport";
+
+interface KernelMiddlewareStack {
+  interrupt: () => void;
+  shutdown: () => void;
+  restart: (onRestarted: ?Function) => void;
+  execute: (
+    code: string,
+    callWatches: boolean,
+    onResults: ResultsCallback
+  ) => void;
+  complete: (code: string, onResults: ResultsCallback) => void;
+  inspect: (
+    code: string,
+    cursorPos: number,
+    onResults: ResultsCallback
+  ) => void;
+}
 
 export default class Kernel {
-  @observable executionState = "loading";
   @observable inspector = { bundle: {} };
   outputStore = new OutputStore();
 
-  kernelSpec: Kernelspec;
-  grammar: atom$Grammar;
-  language: string;
-  displayName: string;
   watchesStore: WatchesStore;
-  watchCallbacks: Array<Function> = [];
   emitter = new Emitter();
   pluginWrapper: HydrogenKernel | null = null;
+  kernel: ZMQKernel | KernelTransport; // XXX(nikita): need explicit ZMQKernel for plugins to typecheck
```

Comment:
Why is there a need for this?
```
$ node
> class A {}
[Function: A]
> class B extends A {}
[Function: B]
> var b = new B();
undefined
> b instanceof B
true
> b instanceof A
```

Reply:
This is not new code I wrote. I just moved from a different file.

Apparently it's needed for flow to work (`npm run-script flow`). Like I said I didn't write it so I haven't looked into this.

Reply:
@lgeiger @BenRussert Does this mean that flow doesn't understand OOP?

Reply:
Looks like I got things to typecheck without this hack, so it should be gone next patchset.